### ISSUE TITLE
World isn't saved before backup.

### ISSUE THREAD

**Issue by daschu117** [spikegrobstein/mcwrapper#27]
When creating backups, your use of [[ ]] to test the return value of a subshell doesn't have the desired effect, causing the world to not be saved before backup. For this kind of test, you don't need brackets and you don't need a subshell. There are other places that you reference this same construct as a TODO item, but the diff output below just contains the active code that causes an issue. This change has the benefit of being simpler and prettier. :-)

``` text
diff --git a/mcwrapper b/mcwrapper
index 8ce46e5..25c81db 100755
--- a/mcwrapper
+++ b/mcwrapper
@@ -543,14 +543,14 @@ function backup_world {
        create_backup_directory

        # stop writing world if we're running
-       if [[ $(check_is_running) ]]; then
+       if check_is_running ; then
                stop_writing_world
        fi

        do_backup

        # start writing world only if we're running
-       if [[ $(check_is_running) ]]; then
+       if check_is_running ; then
                start_writing_world
        fi

```

**Comment by spikegrobstein** [comment]
Hey again!

You should checkout the unstable/1.5 branch. 1.5 is the big cleanup/make friendly release. It has this fix and I believe I've gotten most of the TODO items fixed and also took mcwrapper to the next level.

My original goal was to release 1.5 a lot sooner, but as I noticed my schedule filling up with travel and work, I pushed it back to Oct 1. Had I seen that coming, I would have ported those features into more, minor 1.4 releases.

Please take a look at 1.5 and let me know if you find any issues. I have to sit down and really test the hell out of it before releasing, so if you have your eyes on it, too, you might find things that I miss.

I'm looking at shunit2 for doing unit tests and also considering building this app using `make` (having separate bits of functionality broken out into different files, then stitching together with `make`) to make it easier to manage. It wasn't so bad when this script was like 300 lines, but now that it's closer to 1000, it's getting cumbersome and double and triplechecking the functionality is a pain.

**Comment by daschu117** [comment]
I did double check 1.5 (cuz that would be irresponsible of me to fail to check the next version twice in one day!) for this bit of code before opening this issue and it's still in there. When using the return status of functions and binaries in if statements, a subshell isn't what you're looking for.

This evaluates the text returned by the function, which for check_is_running is empty. It doesn't check the return status at all. 

``` bash
if [[ $(function_name) ]]; then
```

In the course of my testing, I was having check_is_running echo out some text, and as soon as I did that, the conditional basically went "Oh, yes, there is text here. I guess it's true this time" and the backup would finally tell the world to save and toggle live saving.

This actually runs the function (or binary) and uses the exit code in the conditional.

``` bash
if function_name; then
```

It's more or less equivalent to:

``` bash
while true; do
```

Where `true` is the binary that returns true values, resulting in an infinite loop.

But I will checkout 1.5 when I get a chance and I'll take a look at it.

Thanks again :-)

**Comment by spikegrobstein** [comment]
ok, you're correct. it's not in the 1.5 branch. it was in my un-pushed `feature/consolidate_code` branch where I went through and tried to fix those things.

I just pushed that branch... I don't want to merge it into 1.5 until I sufficiently test it.

Thanks for the info. Bash is quite the nuanced language and I'm amazed at how often I learn new things in it. I only just learned about the difference between the above `if` statements a couple weeks ago. Every time I think I'm approaching guru status, something smacks me in the head and I have to re-learn or update my code.
